server/src/main/java/com/cloud/template/TemplateManagerImpl.java

```diff
@@ -366,6 +372,76 @@ public VirtualMachineTemplate registerTemplate(RegisterTemplateCmd cmd) throws U
+            throw new CloudRuntimeException("Please set all the information for a vCenter IP/Name, datacenter, username and password");
+        }
+
+        Map<String, String> params = createParamsForTemplateFromVmwareVmMigration(existingVcenterId,
+                vcenter, datacenterName, username, password, clusterName, sourceHostName, sourceVM);
+
+        HypervisorOutOfBandVMClone cloneResult = vmwareGuru.cloneHypervisorVMOutOfBand(cmd.getHost(), cmd.getVmName(), params);
+        String cloneName = cloneResult.getCloneName();
+        params.put(VmDetailConstants.VMWARE_VM_NAME, cloneName);
+
+        TemplateAdapter adapter = getAdapter(HypervisorType.KVM);
+        TemplateProfile profile = adapter.prepareTemplateFromVmwareMigration(cmd, params, cloneResult);
+        VMTemplateVO vmTemplate = adapter.create(profile);
+
+        if (vmTemplate != null) {
+            CallContext.current().putContextParameter(VirtualMachineTemplate.class, vmTemplate.getUuid());
+            return vmTemplate;
+        } else {
+            throw new CloudRuntimeException("Failed to create a template");
+        }
+    }
+
+    protected Map<String, String> createParamsForTemplateFromVmwareVmMigration(Long existingVcenterId,
+                                                                               String vcenter, String datacenterName,
+                                                                               String username, String password,
+                                                                               String clusterName, String sourceHostName,
+                                                                               String sourceVMName) {
+        Map<String, String> params = new HashMap<>();
+        VmwareDatacenterVO existingDC = null;
+        if (existingVcenterId != null) {
+            existingDC = vmwareDatacenterDao.findById(existingVcenterId);
+            if (existingDC == null) {
+                String err = String.format("Cannot find any existing Vmware DC with ID %s", existingDC);
+                s_logger.error(err);
+                throw new CloudRuntimeException(err);
+            }
+        }
+        params.put(VmDetailConstants.VMWARE_VCENTER, existingDC != null ? existingDC.getVcenterHost() : vcenter);
+        params.put(VmDetailConstants.VMWARE_DATACENTER, existingDC != null ? existingDC.getVmwareDatacenterName() : datacenterName);
+        params.put(VmDetailConstants.VMWARE_VCENTER_USERNAME, existingDC != null ? existingDC.getUser() : username);
+        params.put(VmDetailConstants.VMWARE_VCENTER_PASSWORD, existingDC != null ? existingDC.getPassword() : password);
+        params.put(VmDetailConstants.VMWARE_CLUSTER, clusterName);
+        params.put(VmDetailConstants.VMWARE_HOST, sourceHostName);
+        params.put(VmDetailConstants.VMWARE_VM_NAME, sourceVMName);
```

Comment:
can we use the same variable/constant name ? for example

```suggestion
        params.put(VmDetailConstants.VMWARE_VCENTER_HOST, existingDC != null ? existingDC.getVcenterHost() : vcenterHost);
        params.put(VmDetailConstants.VMWARE_DATACENTER_NAME, existingDC != null ? existingDC.getVmwareDatacenterName() : datacenterName);
        params.put(VmDetailConstants.VMWARE_VCENTER_USERNAME, existingDC != null ? existingDC.getUserName() : username);
        params.put(VmDetailConstants.VMWARE_VCENTER_PASSWORD, existingDC != null ? existingDC.getPassword() : password);
        params.put(VmDetailConstants.VMWARE_CLUSTER_NAME, clusterName);
        params.put(VmDetailConstants.VMWARE_HOST_NAME, sourceHostName);
        params.put(VmDetailConstants.VMWARE_VM_NAME, sourceVMName);
```